subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/support/ClassBytesRepository.kt

```diff
@@ -0,0 +1,191 @@
+    fun classBytesFor(sourceName: String): ByteArray? =
+        classBytesSupplierForSourceName(sourceName)?.let { it() }
+
+    /**
+     * All found class files bytes by Kotlin source name.
+     */
+    fun allClassesBytesBySourceName(): Sequence<Pair<String, ClassBytesSupplier>> =
+        classPathFiles.asSequence()
+            .flatMap { sourceNamesFrom(it) }
+            .mapNotNull { sourceName ->
+                classBytesSupplierForSourceName(sourceName)?.let { sourceName to it }
+            }
+
+    private
+    fun classBytesSupplierForSourceName(sourceName: String): ClassBytesSupplier? =
+        classFilePathCandidatesFor(sourceName)
+            .mapNotNull(::classBytesSupplierForFilePath)
+            .firstOrNull()
+
+    private
+    fun classBytesSupplierForFilePath(classFilePath: String): ClassBytesSupplier? =
+        classBytesIndex.firstNotNullResult { it(classFilePath) }
+
+    private
+    fun sourceNamesFrom(jarOrDir: File): Sequence<String> =
+        when {
+            jarOrDir.isFile -> sourceNamesFromJar(jarOrDir)
+            jarOrDir.isDirectory -> sourceNamesFromDir(jarOrDir)
+            else -> emptySequence()
+        }
+
+    private
+    fun sourceNamesFromJar(jar: File): Sequence<String> =
+        openJarFile(jar).run {
+            entries().asSequence()
+                .filter { it.name.isClassFilePath }
+                .map { kotlinSourceNameOf(it.name) }
+        }
+
+    private
+    fun sourceNamesFromDir(dir: File): Sequence<String> =
+        dir.walkTopDown()
+            .filter { it.name.isClassFilePath }
+            .map { kotlinSourceNameOf(normaliseFileSeparators(it.relativeTo(dir).path)) }
+
+    private
+    fun classBytesIndexFor(jarOrDir: File): ClassBytesIndex =
+        when {
+            jarOrDir.isFile -> jarClassBytesIndexFor(jarOrDir)
+            jarOrDir.isDirectory -> directoryClassBytesIndexFor(jarOrDir)
+            else -> { _ -> null }
+        }
+
+    private
+    fun jarClassBytesIndexFor(jar: File): ClassBytesIndex = { classFilePath ->
+        openJarFile(jar).run {
+            getJarEntry(classFilePath)?.let { jarEntry ->
+                { getInputStream(jarEntry).use { jarInput -> jarInput.readBytes() } }
+            }
+        }
+    }
+
+    private
+    fun directoryClassBytesIndexFor(dir: File): ClassBytesIndex = { classFilePath ->
+        dir.resolve(classFilePath).takeIf { it.isFile }?.let { classFile -> { classFile.readBytes() } }
+    }
+
+    private
+    fun openJarFile(file: File) =
+        openJars.computeIfAbsent(file, ::JarFile)
+
+    override fun close() {
+        openJars.values.forEach(JarFile::close)
+    }
+}
+
+
+private
+val String.isClassFilePath
+    get() = endsWith(classFileExtension)
+        && !endsWith("package-info$classFileExtension")
+        && !matches(compilerGeneratedClassFilePath)
+
+
+private
+const val classFileExtension = ".class"
+
+
+private
+val compilerGeneratedClassFilePath = Regex(".*\\$\\d+\\.class$")
```

Comment:
Probably a stupid question. Is there ever a generated class file with `$` but no number?